src/address.rs

```diff
@@ -40,31 +24,65 @@ impl Address {
     /// creates the zeros. In our case we treat it as "empty"
     /// address, which is then reperesented by zeros.
     pub fn new() -> Address {
-        Address { data: Vec::new() }
+        Address([0; 20])
     }
 
     /// Get raw bytes of the address.
     pub fn as_bytes(&self) -> &[u8] {
-        &self.data
+        &self.0
+    }
+
+    pub fn from_slice(data: &[u8]) -> Result<Address, Error> {
+        ensure!(
+            data.len() == 20,
+            "Address contains exactly 20 bytes but {} found",
+            data.len()
+        );
+        let mut result: [u8; 20] = Default::default();
+        result.copy_from_slice(&data);
+        Ok(Address(result))
     }
 }
 
 impl Default for Address {
     /// Construct a default `Address` filled with zeros.
     fn default() -> Address {
-        Address { data: Vec::new() }
+        Address([0; 20])
+    }
+}
+
+impl Serialize for Address {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for Address {
+    fn deserialize<D>(deserializer: D) -> Result<Address, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        let s = if s.starts_with("0x") { &s[2..] } else { &s };
+
+        hex_str_to_bytes(&s)
+            .and_then(move |bytes| Address::from_slice(&bytes))
+            .map_err(serde::de::Error::custom)
     }
 }
 
 impl From<[u8; 20]> for Address {
     fn from(val: [u8; 20]) -> Address {
-        Address { data: val.to_vec() }
+        Address(val)
     }
 }
 
 impl<'a> From<&'a [u8]> for Address {
     fn from(val: &'a [u8]) -> Address {
-        Address { data: val.to_vec() }
+        Address::from_slice(&val).expect("Unable to convert a slice into a valid Address")
```

Comment:
The trick with `From` is that it mustn't ever fail. You're explicitly breaking that rule here. This trait shouldn't be implemented because not every u8 slice is a valid `Address`.

Reply:
Fixed :+1: